File: src/schemas.ts
/**
 * Typed contracts for the payment path.
 *
 * The split that matters is between what an *agent* proposes and what the
 * platform *decides*. A `PaymentIntent` is a request: it can be produced by a
 * model that just read an attacker-controlled web page, so nothing in it is
 * trusted. A `PaymentDecision` is produced by deterministic code from that
 * intent plus the operator's standing mandates, and it is the only thing that
 * authorizes USDC to move.
 *
 * Note what `PaymentIntent` deliberately cannot express: a disposition. There
 * is no `approved` field for a model to set, no confidence score the engine
 * reads, no `urgent` flag that shortcuts review. The agent's entire influence
 * over the outcome is the counterparty, the amount, and a purpose string —
 * each of which the engine treats as hostile input. A test asserts those
 * fields never appear, so the property survives future edits.
 */

import type { Decimal } from './decimal';

/** Settlement networks. Testnets are the default so a misconfiguration spends play money. */
export type Chain =
  | 'base-sepolia'
  | 'base'
  | 'eth-sepolia'
  | 'ethereum'
  | 'polygon-amoy'
  | 'polygon';

/**
 * How the money moves. `x402` is the HTTP-native pay-per-request handshake
 * Circle's marketplace services speak; `gateway_deposit` funds the
 * Nanopayments gateway; `direct_transfer` is a plain USDC send.
 */
export type PaymentRail = 'x402' | 'gateway_deposit' | 'direct_transfer';

/** What an agent wants to pay, and why. Untrusted by construction. */
export interface PaymentIntent {
  readonly intentId: string;
  /** Wallet address, or the service/origin being paid via x402. */
  readonly counterparty: string;
  readonly amountUsdc: Decimal;
  readonly rail: PaymentRail;
  readonly chain: Chain;
  /** The 402-gated resource, when the rail is x402. */
  readonly resourceUrl?: string;
  /** Agent-supplied, shown to the human reading the approval queue. Never read by the engine. */
  readonly purpose?: string;
  readonly agentId: string;
  readonly requestedAt: string;
  /**
   * Whatever context the agent attached — the tool call, the service listing,
   * the page it was reading. Preserved so a human approving the payment sees
   * what the agent saw, and never consulted by the policy engine.
   */
  readonly context?: Record<string, unknown>;
}

export type Disposition = 'auto_pay' | 'requires_approval' | 'blocked';

/**
 * The deterministic verdict. `reason` is written for a human at 3am; `control`
 * is written for a machine, so the console can group by cause and an operator
 * can tell "over the cap" from "unknown counterparty" without parsing prose.
 */
export interface PaymentDecision {
  readonly intentId: string;
  readonly disposition: Disposition;
  readonly reason: string;
  readonly control: PolicyControl;
  readonly mandateId?: string;
  readonly decidedAt: string;
}

export type PolicyControl =
  | 'kill_switch'
  | 'mainnet_guard'
  | 'amount_unparseable'
  | 'absolute_cap'
  | 'no_mandate'
  | 'mandate_cap'
  | 'window_budget'
  | 'mandated';

/**
 * What actually happened on-chain, or what would have in dry-run.
 *
 * `txHash` plus `explorerUrl` are the difference between a claim and a
 * receipt — and they are literally the proof Circle's rules require: "the
 * agent's Circle wallet address and a clickable block-explorer URL".
 */
export interface PaymentOutcome {
  readonly intentId: string;
  readonly executed: boolean;
  readonly dryRun: boolean;
  readonly detail: string;
  readonly txHash?: string;
  readonly explorerUrl?: string;
  readonly settledAt: string;
  readonly error?: string;
}

export function isAuthorized(decision: PaymentDecision): boolean {
  return decision.disposition === 'auto_pay';
}

File: kronpay/schemas.py
"""
Typed contracts for the payment path.

The split that matters is between what an *agent* proposes and what the
platform *decides*. A `PaymentIntent` is a request — it can be produced by an
LLM that read an attacker-controlled web page, so nothing in it is trusted. A
`PaymentDecision` is produced by deterministic code from the intent plus the
operator's standing mandates, and it is the only thing that can authorize
money leaving a wallet.

Note what a PaymentIntent deliberately cannot express: a disposition. There is
no `allow` field for a model to fill in, no confidence score the engine reads,
no "urgency" that shortcuts approval. The agent's entire influence over the
outcome is the counterparty, the amount, and a purpose string — every one of
which the engine treats as hostile input.
"""
# SPDX-License-Identifier: LicenseRef-PolyForm-Noncommercial-1.0.0
# Copyright (c) 2026 Aditya Kumar, trading as Kronagent · https://kronagent.com

from __future__ import annotations

from datetime import datetime, timezone
from decimal import Decimal
from enum import Enum
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


def utcnow_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class Chain(str, Enum):
    """Settlement networks. Testnet is the default everywhere so that a
    misconfiguration spends play money, not real money."""

    BASE_SEPOLIA = "base-sepolia"
    BASE = "base"
    ETH_SEPOLIA = "eth-sepolia"
    ETHEREUM = "ethereum"

    @property
    def is_testnet(self) -> bool:
        return self in {Chain.BASE_SEPOLIA, Chain.ETH_SEPOLIA}


class PaymentRail(str, Enum):
    """How the money moves. x402 is an HTTP-native pay-per-request handshake;
    a direct transfer is an ordinary USDC send with no request attached."""

    X402 = "x402"
    DIRECT_TRANSFER = "direct_transfer"


class PaymentIntent(BaseModel):
    """What an agent wants to pay, and why. Untrusted by construction."""

    model_config = ConfigDict(frozen=True)

    intent_id: str
    counterparty: str            # wallet address or x402 resource origin
    amount_usdc: Decimal         # USDC, not micro-units — six decimal places
    rail: PaymentRail = PaymentRail.X402
    chain: Chain = Chain.BASE_SEPOLIA
    resource_url: str = ""       # the 402-gated URL, when rail is x402
    purpose: str = ""            # agent-supplied, for the human reading the queue
    requested_at: str = Field(default_factory=utcnow_iso)
    agent_id: str = "default"
    # Free-form context the agent attached. Never read by the policy engine —
    # it exists so a human approving the payment can see what the agent saw.
    context: dict[str, Any] = Field(default_factory=dict)

    @field_validator("amount_usdc")
    @classmethod
    def _positive(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("amount_usdc must be greater than zero")
        return v

    def as_audit_payload(self) -> dict[str, Any]:
        return {
            "intent_id": self.intent_id,
            "counterparty": self.counterparty,
            "amount_usdc": str(self.amount_usdc),
            "rail": self.rail.value,
            "chain": self.chain.value,
            "resource_url": self.resource_url,
            "purpose": self.purpose,
            "agent_id": self.agent_id,
        }


Disposition = Literal["auto_pay", "requires_approval", "blocked"]


class PaymentDecision(BaseModel):
    """The deterministic verdict. `reason` is written to be read by a human at
    3am, and by a judge reading an audit export six months later."""

    model_config = ConfigDict(frozen=True)

    intent_id: str
    disposition: Disposition
    reason: str
    # Which control produced this verdict — so an operator can tell "over the
    # per-payment cap" from "counterparty not mandated" without parsing prose.
    control: str
    mandate_id: Optional[str] = None
    decided_at: str = Field(default_factory=utcnow_iso)

    @property
    def authorizes_payment(self) -> bool:
        return self.disposition == "auto_pay"


class PaymentOutcome(BaseModel):
    """What actually happened on-chain, or what would have happened in dry-run.

    `tx_hash` is the difference between a claim and a receipt: it is the thing
    a judge, an auditor or a finance team can independently verify.
    """

    intent_id: str
    executed: bool
    dry_run: bool
    detail: str
    tx_hash: str = ""
    explorer_url: str = ""
    settled_at: str = Field(default_factory=utcnow_iso)
    error: str = ""
